pkg/storage/batch.go

```diff
@@ -751,6 +774,77 @@ func fetchLazyChunks(ctx context.Context, s chunk.SchemaConfig, chunks []*LazyCh
 	return nil
 }
 
+func postFetchFilter(ctx context.Context, chunks []chunk.Chunk, filterer PostFetcherChunkFilterer, schema chunk.SchemaConfig) ([]chunk.Chunk, error) {
+	if len(chunks) == 0 {
+		return chunks, nil
+	}
+	postFilterChunkLen := 0
+	log, ctx := spanlogger.New(ctx, "Batch.ParallelPostFetchFilter")
+	log.Span.LogFields(otlog.Int("chunks", len(chunks)))
+	defer func() {
+		log.Span.LogFields(otlog.Int("postFilterChunkLen", postFilterChunkLen))
+		log.Span.Finish()
+	}()
+
+	logSelector, err := logql.ParseLogSelector(filterer.Request().Selector, true)
+	if err != nil {
+		return nil, err
+	}
+	if !logSelector.HasFilter() {
+		return chunks, nil
+	}
+	pipeline, err := logSelector.Pipeline()
+	if err != nil {
+		return nil, err
+	}
+	//
+	streamPipeline := pipeline.ForStream(chunks[0].Metric.WithoutLabels(labels.MetricName))
+	result := make([]chunk.Chunk, 0)
+
+	for _, cnk := range chunks {
+		chunkData := cnk.Data
+		lokiChunk := chunkData.(*chunkenc.Facade).LokiChunk()
+		blocks := lokiChunk.Blocks(filterer.Request().Start, filterer.Request().End)
+		if len(blocks) == 0 {
+			continue
+		}
+		postFilterChunkData := chunkenc.NewMemChunk(filterer.Encoding(), filterer.HeadBlockFmt(), cnk.Data.Size(), cnk.Data.Size())
```

Comment:
For encoding you can use `lokiChunk.Encoding()` and `chunkenc.OrderedHeadBlockFmt` for headblock format here.

Reply:
done.thanks